File: src/main/java/net/orandja/vw/ExtraSaveData.java
package net.orandja.vw;

import com.google.common.collect.Maps;
import net.minecraft.SharedConstants;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtIo;
import net.minecraft.util.registry.RegistryKey;
import net.minecraft.world.World;
import net.orandja.mcutils.NBTUtils;

import java.io.*;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;

public interface ExtraSaveData {

    interface ExtraSaveDataSupplier {
        NbtCompound supply();
    }

    private static void readNbt() {
        if(file.exists()) {
            PushbackInputStream PBStream = null;
            DataInputStream DIStream = null;
            try {
                PBStream = new PushbackInputStream(new FileInputStream(file), 2);
                DIStream = new DataInputStream(PBStream);
                loadedCompound.set(NbtIo.readCompressed(PBStream));
            } catch(Exception e) {
                e.printStackTrace();
                loadedCompound.set(new NbtCompound());
            } finally {
                try {
                    if(PBStream != null) {
                        PBStream.close();
                    }
                    if(DIStream != null) {
                        DIStream.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    static void onLoad(String name, Consumer<NbtCompound> consumer) {
        LOAD.put(name, consumer);
        if(loadedCompound.get() != null) {
            NbtCompound dataCompound = loadedCompound.get().getCompound("data");
            if(dataCompound.contains(name)) {
                consumer.accept(dataCompound.getCompound(name));
            }
        }
    }

    static void onSave(String name, ExtraSaveDataSupplier supplier) {
        SAVE.put(name, supplier);
    }

    Map<String, Consumer<NbtCompound>> LOAD = Maps.newHashMap();
    Map<String, ExtraSaveDataSupplier> SAVE = Maps.newHashMap();
    File file = new File("config/vw-extradata.dat");

    AtomicReference<NbtCompound> loadedCompound = new AtomicReference<>(null);

    default void loadExtraSaveData(RegistryKey<World> registryKey) {
        if(registryKey != World.OVERWORLD) {
            return;
        }

        if(loadedCompound.get() == null) {
            readNbt();
        }
        NbtCompound dataCompound = loadedCompound.get().getCompound("data");
        LOAD.forEach((name, consumer) -> {
            if(dataCompound.contains(name))
                consumer.accept(dataCompound.getCompound(name));
        });
    }

    default void saveExtraSaveData(RegistryKey<World> registryKey) {
        if(registryKey != World.OVERWORLD) {
            return;
        }

        NbtCompound mainCompound = new NbtCompound();
        mainCompound.put("data", NBTUtils.createBlankCompound(it -> {
            SAVE.forEach((name, supplier) -> it.put(name, supplier.supply()));
        }));
        mainCompound.putInt("DataVersion", SharedConstants.getGameVersion().getWorldVersion());

        try {
            NbtIo.writeCompressed(mainCompound, file);
        } catch (IOException e) {
            e.printStackTrace();
        }

        loadedCompound.set(mainCompound);
    }
}

File: src/main/kotlin/net/orandja/vw/ExtraSaveData.kt
@file:Suppress("unused")

package net.orandja.vw

import net.minecraft.SharedConstants
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtIo
import net.minecraft.util.registry.RegistryKey
import net.minecraft.world.World
import java.io.*

@SuppressWarnings("unused")
interface ExtraSaveData {
    companion object {

        val CLEAR: HashMap<String, () -> Unit> = HashMap()
        val LOAD: HashMap<String, (NbtCompound) -> Unit> = HashMap()
        val SAVE: HashMap<String, () -> NbtCompound> = HashMap()

        private val file = File("config/vw-extradata.dat")
        var loadedCompound: NbtCompound? = null

        private fun readNbt() {
            if (File("config/vw-extradata.dat").exists()) {
                var pushbackInputStream: PushbackInputStream? = null
                try {
                    pushbackInputStream = PushbackInputStream(FileInputStream(file), 2)
                    val dataInputStream = DataInputStream(pushbackInputStream)
                    loadedCompound = NbtIo.readCompressed(pushbackInputStream)
                    dataInputStream.close()
                    pushbackInputStream.close()
                } catch (e: Exception) {
                    pushbackInputStream?.close()
                    e.printStackTrace()
                    loadedCompound = NbtCompound()
                }
            } else {
                loadedCompound = NbtCompound()
            }
        }

        fun onLoad(name: String, consumer: (NbtCompound) -> Unit) {
            LOAD[name] = consumer
            if (loadedCompound != null) {
                val dataCompound = loadedCompound!!.getCompound("data")
                if (dataCompound.contains(name))
                    consumer(dataCompound.getCompound(name))
            }
        }

        fun clear() {
            CLEAR.forEach { entry -> entry.value.invoke() }
        }

        fun onClear(name: String, consumer: () -> Unit) {
            CLEAR[name] = consumer
        }

        fun onSave(name: String, consumer: () -> NbtCompound) {
            SAVE[name] = consumer
        }

    }
    fun load(registryKey: RegistryKey<World>) {
        if (registryKey !== World.OVERWORLD) {
            return
        }

        if (loadedCompound == null) {
            readNbt()
        }
        val dataCompound = loadedCompound!!.getCompound("data")
        LOAD.forEach {
            if (dataCompound.contains(it.key))
                it.value(dataCompound.getCompound(it.key))
        }
    }

    fun save(registryKey: RegistryKey<World>) {
        if (registryKey !== World.OVERWORLD) {
            return
        }
        val mainCompound = NbtCompound()
        mainCompound.put("data", NbtCompound().apply {
            SAVE.forEach { put(it.key, it.value()) }
        })
        @Suppress("DEPRECATION")
        mainCompound.putInt("DataVersion", SharedConstants.getGameVersion().worldVersion)

        try {
            NbtIo.writeCompressed(mainCompound, file)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        loadedCompound = mainCompound
    }
}